### PR TITLE
gh-124688: _decimal: Get a module state from ctx objects for performance

### PR DESCRIPTION
**Pull request by neonene**
* Issue: gh-124688

### CODE REVIEW

**Comment by vstinner** [comment]
Do you have a benchmark showing the effect of this change?

**Comment by neonene** [comment]
`_dec_mpd_radix()` is added (818c24fd2c96d8e1494528397a72c14352be5bd6) so as not to be passed a different object, not for performance.

**Comment by neonene** [comment]
On release builds, the result was as fast as the global state access: https://github.com/python/cpython/pull/123100#issuecomment-2298566533 (`PR with alternatives`: 1.08x  )

**Comment by neonene** [comment]
> Would you mind to share your results (before/after, difference)?
Wait please.

**Comment by neonene** [comment]
Picked a few `telco` results on Windows PGO builds:

a) main with d69529d31ccd1510843cfac1ab53bb8cb027541f restored

```
9.68 ms +- 0.17 ms -> 8.94 ms +- 0.15 ms: 1.08x faster  # PR (ctx object)
9.68 ms +- 0.17 ms -> 8.76 ms +- 0.11 ms: 1.10x faster  # global state access
```

b) (a) with some patches

```
8.99 ms +- 0.11 ms -> 8.49 ms +- 0.16 ms: 1.06x faster  # PR (ctx object)
8.99 ms +- 0.11 ms -> 8.44 ms +- 0.20 ms: 1.07x faster  # global state access
```

c) (b) with 1 diff in `PyType_GetBaseByToken()`

```
8.91 ms +- 0.21 ms -> 8.78 ms +- 0.16 ms: 1.02x faster  # PR (ctx object)
8.91 ms +- 0.21 ms -> 8.25 ms +- 0.17 ms: 1.08x faster  # global state access
```

**Comment by vstinner** [comment]
Sorry, I don't know how to read your 6 benchmarks. I don't understand what you compared. I don't know what are "some patches" nor why you mention PyType_GetBaseByToken(), it doesn't help me to compare *this* PR.

So I just compared the main branch to this PR:

```
Mean +- std dev: [ref] 8.18 ms +- 0.28 ms -> [change] 7.53 ms +- 0.18 ms: 1.09x faster
```

Nice! 1.09x faster is worth it.

File: Modules/_decimal/_decimal.c [hunk 1]
```diff
@@ -210,6 +212,15 @@ typedef struct {
 #define CTX(v) (&((PyDecContextObject *)v)->ctx)
 #define CtxCaps(v) (((PyDecContextObject *)v)->capitals)
 
+static inline decimal_state *
+get_module_state_from_ctx(PyObject *v)
+{
+    assert(PyType_GetBaseByToken(Py_TYPE(v), &context_spec, NULL) == 1);
+    void *state = ((PyDecContextObject *)v)->modstate;
+    assert(state != NULL);
+    return (decimal_state *)state;
```
Comment (vstinner) [inline Modules/_decimal/_decimal.c]:
```suggestion
    decimal_state *state = ((PyDecContextObject *)v)->modstate;
    assert(state != NULL);
    return state;
```

**Comment by picnixz** [comment]
I added the skip news since Victor enabled auto-merge. If a NEWS entry is really needed, we can add it later in a separate commit though.

**Comment by vstinner** [comment]
Merged, thanks @neonene!

**Pull request merged**